### PR TITLE
Experimenting with spdlog

### PR DESCRIPTION
**Pull request by mjcarroll**

### CODE REVIEW

**Comment by j-rivero** [comment]
Unsure if the comment is out of the scope of this work, please ignore it if it is.

The current implementation seems to support different "sinks" (destinations for the logging) in the sense that I can see the fileSink and consoleSink in the implementation of the `ConsoleNew::Implementation`. Even if the class naming and file naming is a bit confuse since we use "Console" all over the place.

I think that supported this variety of "sinks" is really a great approach so we can extend the logging capabilities in the future. Concretely I'm looking into experimenting with [OpenTelemetry](https://opentelemetry.io/) for the whole Gazebo and one of the options that this spdlog support can provide is the use of an ["OpenTelemetry sink"](https://github.com/open-telemetry/opentelemetry-cpp-contrib/tree/main/instrumentation/spdlog). The creation of it seems trivial if we can support multiple sinks somehow:

```c++
#include <opentelemetry/instrumentation/spdlog/sink.h>

auto otel_sink = std::make_shared<spdlog::sinks::opentelemetry_sink_mt>();
otel_sink->set_level(spdlog::level::info);
auto logger = spdlog::logger("OTelLogger", otel_sink);
```

**Comment by mjcarroll** [comment]
> The current implementation seems to support different "sinks" (destinations for the logging) in the sense that I can see the fileSink and consoleSink in the implementation of the ConsoleNew::Implementation. Even if the class naming and file naming is a bit confuse since we use "Console" all over the place.

That is correct.  I used `ConsoleNew` to keep the name and behavior consistent with the current `Console` behavior.  The current `Console` also sinks to file and stdout/stderr console streams.

In this implementation, I wanted to make sure that we can surface the underlying spdlog mechanisms so you could (for instance) add an OpenTelemetry sink to Console.

**Comment by azeey** [comment]
Would it make sense to move this into gz-utils? We have an implementation of `Console`  in sdformat which could be replaced with this. I was also working on https://github.com/gazebosim/gz-math/pull/613 which needs to enforce an invariant. Since we don't use exceptions, this will probably resort to printing an error message, which we do in many places in gz-math. It would be nice if we had proper logging there as well.

**Comment by azeey** [comment]
@mjcarroll and I discussed putting a lightweight wrapper around `spdlog` in gz-utils which would allow us to use it in libraries such as `gz-math` or `sdformat`. We would still have the higher level API, aka `common::Console` here which will have more functionality. This would provide a way for downstream applications like gz-sim capture and redirect console output from `gz-math` or `sdformat` and display them however they want.

**Comment by azeey** [comment]
Rebuilding gz-utils [Build Status](https://build.osrfoundation.org/job/gz-utils3-debbuilder/518/)

File: package.xml [hunk 1]
```diff
@@ -18,6 +18,8 @@
   <depend>gz-utils3</depend>
   <depend>libfreeimage-dev</depend>
   <depend>libgdal-dev</depend>
+  <depend>libgts</depend>
```
Comment (azeey) [inline package.xml:21]: I think `libgts` was removed in https://github.com/gazebosim/gz-common/commit/1f527012d21fd5afde75de8b49e0776e08f3cd6c

**Comment by iche033** [comment]
update migration guide?

File: package.xml (hunk 1 above)
Comment (caguero) [reply package.xml]: Removed in [f1db533](https://github.com/gazebosim/gz-common/pull/615/commits/f1db533d430784d6fa63fbc0988a355f0ffa05a7).

**Comment by iche033** [comment]
I tested this with gz-sim's `main` branch and I noticed a couple of issues. 

1) This first one is minor - there are now empty lines between console log entries

```
[2024-08-26 21:48:33.501] [gz] [info] [SceneBroadcaster.cc:634] Serving scene information on [/world/shapes/scene/info]

[2024-08-26 21:48:33.502] [gz] [info] [SceneBroadcaster.cc:643] Serving graph information on [/world/shapes/scene/graph]

[2024-08-26 21:48:33.502] [gz] [info] [SceneBroadcaster.cc:654] Serving full state on [/world/shapes/state]
```

2) I get a crash if I try to close the GUI window:

```
terminate called after throwing an instance of 'std::system_error'
...
...
#13   Object "/lib/x86_64-linux-gnu/libspdlog.so.1.12", at 0x7fa4dfdb3400, in spdlog::details::registry::~registry()
#12   Object "/lib/x86_64-linux-gnu/libspdlog.so.1.12", at 0x7fa4dfde771b, in spdlog::details::periodic_worker::~periodic_worker()
```

**Comment by caguero** [comment]
> 27633b7

Updated in [27633b7](https://github.com/gazebosim/gz-common/pull/615/commits/27633b7f823ff5d3b679c3d695dfc201d6591658).

**Comment by caguero** [comment]
> update migration guide?

Updated in [27633b7](https://github.com/gazebosim/gz-common/pull/615/commits/27633b7f823ff5d3b679c3d695dfc201d6591658)

File: src/Console_TEST.cc [hunk 2]
```diff
@@ -89,7 +90,7 @@ TEST_F(Console_TEST, NoInitAndLog)
   std::string path;
   EXPECT_TRUE(common::env(GZ_HOMEDIR, path));
   path = common::joinPaths(path, logPath);
-  EXPECT_TRUE(common::removeAll(path));
+  common::removeAll(path);
```
Comment (azeey) [inline src/Console_TEST.cc:93]: I'm curious why this `EXPECT_TRUE` was removed.

File: src/Console_TEST.cc [hunk 3]
```diff
@@ -470,13 +489,14 @@ TEST_F(Console_TEST, ColorErr)
 /// \brief Test Console::Verbosity
 TEST_F(Console_TEST, Verbosity)
 {
-  EXPECT_EQ(common::Console::Verbosity(), 1);
```
Comment (azeey) [inline src/Console_TEST.cc:473]: It looks like this was testing the default verbosity? Is this now different?

File: test/integration/console.cc [hunk 4]
```diff
@@ -37,7 +37,7 @@ TEST(Console_TEST, LogInitAfterConsoleOut)
   gz::common::Console::SetVerbosity(4);
 
   // Before console is initialized, logs go to default file location
-  gzerr << "Error before logging initialized" << std::endl;
+  gzlog << "Error before logging initialized" << std::endl;
```
Comment (azeey) [inline test/integration/console.cc:40]: Does `gzerr` not log to the default file location?

File: src/Console_TEST.cc [hunk 5]
```diff
@@ -505,14 +525,16 @@ TEST_F(Console_TEST, Prefix)
...
+
   // Get the logged content
   std::string logContent = GetLogContent(logPath);
 
   // Check
-  EXPECT_TRUE(logContent.find("**test** [Err]") != std::string::npos);
-  EXPECT_TRUE(logContent.find("**test** [Wrn]") != std::string::npos);
-  EXPECT_TRUE(logContent.find("**test** [Msg]") != std::string::npos);
-  EXPECT_TRUE(logContent.find("**test** [Dbg]") != std::string::npos);
+  EXPECT_TRUE(logContent.find("**test** error") != std::string::npos);
```
Comment (azeey) [inline src/Console_TEST.cc:534]: Is this a behavior change on the formatting of the logs? Should this be included in the migration guide?

File: src/Console.cc [hunk 6]
```diff
@@ -14,223 +14,74 @@
...
+  spdlog::level::level_enum _logLevel, bool _fileInitialize)
+  : severity(_logLevel),
+    sourceLocation(_file, _line, "")
 {
+  // Use default initialization if needed.
+  if (_fileInitialize && !Console::initialized)
+  {
+    Console::Init(".gz", "auto_default.log");
+    // Allow gzlog to pass through.
+    Console::SetVerbosity(5);
```
Comment (azeey) [inline src/Console.cc:48]: Do we want to set the verbosity to 5 here? On line 66, I see `Console::verbosity = 1`.

**Comment by iche033** [comment]
Added `spdlog` dependency to:
* homebrew: https://github.com/osrf/homebrew-simulation/pull/2747
* gz-common6-release: https://github.com/gazebo-release/gz-common6-release/pull/5

windows has `spdlog` dependency already (https://github.com/gazebo-tooling/release-tools/pull/1163)

**Comment by caguero** [comment]
> I tested this with gz-sim's `main` branch and I noticed a couple of issues.
> 
> 1. This first one is minor - there are now empty lines between console log entries
> 
> ```
> [2024-08-26 21:48:33.501] [gz] [info] [SceneBroadcaster.cc:634] Serving scene information on [/world/shapes/scene/info]
> 
> [2024-08-26 21:48:33.502] [gz] [info] [SceneBroadcaster.cc:643] Serving graph information on [/world/shapes/scene/graph]
> 
> [2024-08-26 21:48:33.502] [gz] [info] [SceneBroadcaster.cc:654] Serving full state on [/world/shapes/state]
> ```
> 
> 2. I get a crash if I try to close the GUI window:
> 
> ```
> terminate called after throwing an instance of 'std::system_error'
> ...
> ...
> #13   Object "/lib/x86_64-linux-gnu/libspdlog.so.1.12", at 0x7fa4dfdb3400, in spdlog::details::registry::~registry()
> #12   Object "/lib/x86_64-linux-gnu/libspdlog.so.1.12", at 0x7fa4dfde771b, in spdlog::details::periodic_worker::~periodic_worker()
> ```

Could you try with this PR? https://github.com/gazebosim/gz-utils/pull/139

I think we should use a period flush for efficiency but we might be doing something wrong and I suspect it's related with the use of  `static`. This could be a workaround until we learn more about the issue.

**Comment by iche033** [comment]
> Could you try with this PR? [gazebosim/gz-utils#139](https://github.com/gazebosim/gz-utils/pull/139)

tested and confirmed it fixes the issues.

File: src/Console.cc (hunk 6 above)
Comment (azeey) [reply src/Console.cc]: Fixed in https://github.com/gazebosim/gz-common/pull/615/commits/f3d168cfdc7eb4a8575edc4dfbf24cd8a98a5c52

File: test/integration/console.cc (hunk 4 above)
Comment (azeey) [reply test/integration/console.cc]: Reverted in https://github.com/gazebosim/gz-common/pull/615/commits/f3d168cfdc7eb4a8575edc4dfbf24cd8a98a5c52

File: testing/src/AutoLogFixture_TEST.cc [hunk 7]
```diff
@@ -28,7 +28,8 @@ TEST_F(AutoLogFixture, AutoLogFixture)
   EXPECT_FALSE(this->FullLogPath().empty());
   EXPECT_TRUE(this->LogContent().empty());
 
-  Console::SetVerbosity(0);
+  Console::SetVerbosity(4);
```
[the remark this answers is not in the document]
Comment (azeey) [reply testing/src/AutoLogFixture_TEST.cc]: Reverted in https://github.com/gazebosim/gz-common/pull/615/commits/f3d168cfdc7eb4a8575edc4dfbf24cd8a98a5c52

[the remark this answers is not in the document]
Comment (caguero) [reply testing/src/AutoLogFixture_TEST.cc]: Restored to `0` in [f3d168c](https://github.com/gazebosim/gz-common/pull/615/commits/f3d168cfdc7eb4a8575edc4dfbf24cd8a98a5c52)

File: src/Console_TEST.cc (hunk 3 above)
Comment (caguero) [reply src/Console_TEST.cc:473]: This line has been restored in [f3d168c](https://github.com/gazebosim/gz-common/pull/615/commits/f3d168cfdc7eb4a8575edc4dfbf24cd8a98a5c52).

File: test/integration/console.cc (hunk 4 above)
Comment (caguero) [reply test/integration/console.cc]: gzerr restored in [f3d168c](https://github.com/gazebosim/gz-common/pull/615/commits/f3d168cfdc7eb4a8575edc4dfbf24cd8a98a5c52).

File: src/Console.cc (hunk 6 above)
Comment (caguero) [reply src/Console.cc]: Removed this changed in [f3d168c](https://github.com/gazebosim/gz-common/pull/615/commits/f3d168cfdc7eb4a8575edc4dfbf24cd8a98a5c52).

File: src/Console_TEST.cc (hunk 2 above)
Comment (caguero) [reply src/Console_TEST.cc:96]: Added a note in [ab269e8](https://github.com/gazebosim/gz-common/pull/615/commits/ab269e8eee1c0bc837f3940f6dda25f66842ba69).

**Review by azeey — APPROVED** [review]

**Comment by caguero** [comment]
@mergifyio backport gz-common6

**Pull request merged**